cirq-google/cirq_google/devices/grid_device.py

```diff
@@ -77,6 +83,91 @@ def _validate_device_specification(proto: v2.device_pb2.DeviceSpecification) ->
+
+    # TODO(#5050) Describe how to add/remove gates.
+
+    for gate_spec in proto.valid_gates:
+        gate_name = gate_spec.WhichOneof('gate')
+        cirq_gates: List[Union[Type[cirq.Gate], cirq.Gate, cirq.GateFamily]] = []
+
+        if gate_name == 'syc':
+            cirq_gates = [ops.SYC]
+            fsim_gates.append(ops.SYC)
+        elif gate_name == 'sqrt_iswap':
+            cirq_gates = [cirq.SQRT_ISWAP]
+            fsim_gates.append(cirq.SQRT_ISWAP)
+        elif gate_name == 'sqrt_iswap_inv':
+            cirq_gates = [cirq.SQRT_ISWAP_INV]
+            fsim_gates.append(cirq.SQRT_ISWAP_INV)
+        elif gate_name == 'cz':
+            cirq_gates = [cirq.CZ]
+            fsim_gates.append(cirq.CZ)
+        elif gate_name == 'phased_xz':
+            cirq_gates = [
+                cirq.PhasedXZGate,
+                cirq.XPowGate,
+                cirq.YPowGate,
+                cirq.ZPowGate,
+                cirq.PhasedXPowGate,
+            ]
+        elif gate_name == 'virtual_zpow':
+            cirq_gates = [cirq.GateFamily(cirq.ZPowGate, tags_to_ignore=[ops.PhysicalZTag()])]
+        elif gate_name == 'physical_zpow':
+            cirq_gates = [cirq.GateFamily(cirq.ZPowGate, tags_to_accept=[ops.PhysicalZTag()])]
+        elif gate_name == 'coupler_pulse':
+            cirq_gates = [experimental_ops.CouplerPulse]
+        elif gate_name == 'meas':
+            cirq_gates = [cirq.MeasurementGate]
+        elif gate_name == 'wait':
+            cirq_gates = [cirq.WaitGate]
+        else:
+            # coverage: ignore
+            warnings.warn(
+                f"The DeviceSpecification contains the gate '{gate_name}' which is not recognized"
+                " by Cirq and will be ignored. This may be due to an out-of-date Cirq version.",
+                UserWarning,
+            )
+            continue
+
+        gates_list.extend(cirq_gates)
+        for g in cirq_gates:
+            if not isinstance(g, cirq.GateFamily):
+                g = cirq.GateFamily(g)
+            gate_durations[g] = cirq.Duration(picos=gate_spec.gate_duration_picos)
+
+    if fsim_gates:
+        gates_list.append(ops.FSimGateFamily(gates_to_accept=fsim_gates))
+
+    # TODO(#4833) Add identity gate support
+    # TODO(#5050) Add GlobalPhaseGate support
+
+    return cirq.Gateset(*gates_list), gate_durations
+
+
+def _build_compilation_target_gatesets(
+    gateset: cirq.Gateset,
+) -> Sequence[cirq.CompilationTargetGateset]:
+    """Detects compilation target gatesets based on what gates are inside the gateset."""
+
+    target_gatesets: List[cirq.CompilationTargetGateset] = []
+    if cirq.CZ in gateset:
+        target_gatesets.append(cirq.CZTargetGateset())
+    if ops.SYC in gateset:
+        target_gatesets.append(transformers.SycamoreTargetGateset())
+    if cirq.SQRT_ISWAP in gateset:
+        target_gatesets.append(
+            cirq.SqrtIswapTargetGateset(use_sqrt_iswap_inv=cirq.SQRT_ISWAP_INV in gateset)
+        )
```

Comment:
As we discussed offline, automatically ignoring the problematic gates on Google devices (WaitGate, ZPowGates with a PhysicalZ) in compilation is tricky because `tags_to_ignore` need to be appended to all contexts, and we'll document the behavior and suggest users to specify their own `tags_to_ignore` in their transformer context. We can pursue a more complete solution in the future when the uses cases become common, by creating subclasses of target gatesets in core inside cirq-google.

Reply:
After further offline discussion, we decided to make a fix for ZPowGates now, but will do this in a series of follow-up PRs instead.

Reply:
Can you add a TODO to the code saying that we should replace cirq-core versions with cirq-google specific implementations which correctly deal with cases like physical Zs etc. ?